File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/opengl/renderer/GLGenericRenderer.java
package br.com.insanitech.spritekit.opengl.renderer;

import android.opengl.GLSurfaceView;
import br.com.insanitech.spritekit.opengl.context.GLContextFactory;
import br.com.insanitech.spritekit.opengl.model.GLColor;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

/**
 * Created by anderson on 7/2/15.
 */
public class GLGenericRenderer implements GLSurfaceView.Renderer {
    private GLRenderer usedRenderer;

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        if (usedRenderer != null) {
            usedRenderer.onSurfaceCreated(gl, config);
        }
    }

    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {
        if (usedRenderer != null) {
            usedRenderer.onSurfaceChanged(gl, width, height);
        }
    }

    @Override
    public void onDrawFrame(GL10 gl) {
        if (usedRenderer != null) {
            usedRenderer.onDrawFrame(gl);
        }
    }

    public void setDrawer(GLRenderer.GLDrawer drawer) {
        if (usedRenderer != null) {
            usedRenderer.setDrawer(drawer);
        }
    }

    public void setGLVersion(float version) {
        switch ((int)version) {
            case (int) GLContextFactory.GLVersion.GL10:
                if (version == GLContextFactory.GLVersion.GL10) {
                    usedRenderer = new GL10Renderer();
                } else if (version >= GLContextFactory.GLVersion.GL11) { // works for GL11 and GL14
                    usedRenderer = new GL11Renderer();
                }
                break;


            case (int) GLContextFactory.GLVersion.GL20:
                usedRenderer = new GL20Renderer();
                break;

            case (int) GLContextFactory.GLVersion.GL30:
                usedRenderer = new GL30Renderer();
                break;
        }
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/opengl/renderer/GLGenericRenderer.kt
package br.com.insanitech.spritekit.opengl.renderer

import android.opengl.GLSurfaceView
import br.com.insanitech.spritekit.opengl.context.GLContextFactory

import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

/**
 * Created by anderson on 7/2/15.
 */
internal class GLGenericRenderer(private val drawer: GLRenderer.GLDrawer) : GLSurfaceView.Renderer {
    private lateinit var usedRenderer: GLRenderer

    override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        this.usedRenderer.onSurfaceCreated(gl, config)
    }

    override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
        this.usedRenderer.onSurfaceChanged(gl, width, height)
    }

    override fun onDrawFrame(gl: GL10) {
        this.usedRenderer.onDrawFrame(gl)
    }

    fun setGLVersion(version: Float) {
        when (version.toInt()) {
            GLContextFactory.GLVersion.GL10.toInt() -> when {
                version == GLContextFactory.GLVersion.GL10 ->
                    this.usedRenderer = GL10Renderer(this.drawer)

                version >= GLContextFactory.GLVersion.GL11 -> // works for GL11 and GL14
                    this.usedRenderer = GL11Renderer(this.drawer)
            }


            GLContextFactory.GLVersion.GL20.toInt() ->
                this.usedRenderer = GL20Renderer(this.drawer)

            GLContextFactory.GLVersion.GL30.toInt() ->
                this.usedRenderer = GL30Renderer(this.drawer)
        }
    }
}
